// nodedata/core/decorators/processControl.js
"use strict";
const utils_1 = require("../metadata/utils");
const Utils = require("../utils");
const decorators_1 = require('../constants/decorators');
const decorator_type_1 = require('../enums/decorator-type');
const IProcessControlService_1 = require('./interfaces/IProcessControlService');
const Enumerable = require('linq');
const workerAssociation_1 = require('./workerAssociation');
const di_1 = require('../../di');
const Q = require('q');
//This function is a private function and not exposed as an attribute
//process type=1 process_start
//process type=2 processend
//process type=3 processstart and end
function preProcessHandler(params, target, propertyKey, descriptor, originalMethod, type) {
    return function () {
        console.log("preProcessHandler", params);
        var meta = utils_1.MetaUtils.getMetaData(target, type, propertyKey);
        var targetObjectId;
        if (params.indexofArgumentForTargetObjectId != undefined) {
            targetObjectId = arguments[params.indexofArgumentForTargetObjectId];
        }
        if (params.indexofArgumentForTargetObject != undefined) {
            targetObjectId = arguments[params.indexofArgumentForTargetObject] && arguments[params.indexofArgumentForTargetObject]._id;
        }
        var services = utils_1.MetaUtils.getMetaDataForDecorators([decorators_1.Decorators.SERVICE]);
        var procService = Enumerable.from(services).where(x => x.metadata[0].params.serviceName == IProcessControlService_1.processControlServiceName).select(x => x.metadata[0]).firstOrDefault();
        if (!procService)
            return originalMethod.apply(this, arguments);
        var processControlService = di_1.Container.resolve(procService.params.target);
        var targetServices = utils_1.MetaUtils.getMetaData(target, decorators_1.Decorators.SERVICE);
        var serviceName = targetServices[0].params.serviceName;
        if (processControlService && meta) {
            if (type == decorators_1.Decorators.PROCESS_START || type == decorators_1.Decorators.PROCESS_START_AND_END) {
                //preprocess
                let argsObj = Utils.getMethodArgs(originalMethod, arguments);
                if (params.executeInDistributedWorker) {
                    // Parent - processcontrol management executing with worker 
                    console.log('process control initializing started');
                    let workerParams = workerAssociation_1.executeWorkerHandler({}, target, propertyKey, originalMethod, decorators_1.Decorators.WORKER, true).apply(this, arguments);
                    return processControlService.initialize(serviceName, propertyKey, targetObjectId, params, workerParams).then((sucess) => {
                        // var taskInfo = executeWorkerHandler({}, target, propertyKey, originalMethod, Decorators.WORKER).apply(this, arguments);
                        console.log('process control initialized successfully');
                        return processControlService.sendResponse(sucess);
                    });
                }
                else if (params.executeInWorker && !utils_1.MetaUtils.childProcessId && Utils.config().Config.isMultiThreaded) {
                    // Parent - processcontrol management executing with worker 
                    console.log('process control initializing started');
                    return processControlService.initialize(serviceName, propertyKey, targetObjectId, params).then((sucess) => {
                        var taskInfo = workerAssociation_1.executeWorkerHandler({}, target, propertyKey, originalMethod, decorators_1.Decorators.WORKER).apply(this, arguments);
                        console.log('process control initialized successfully');
                        return processControlService.sendResponse(sucess);
                    });
                }
                else {
                    let initialize;
                    if (params.executeInWorker && utils_1.MetaUtils.childProcessId) {
                        // already initialized in parent process
                        initialize = Q.when(true);
                        console.log('process control already initialized');
                    }
                    else {
                        console.log('process control initialized started');
                        initialize = processControlService.initialize(serviceName, propertyKey, targetObjectId, params, argsObj);
                    }
                    return initialize.then(res => {
                        return processControlService.startProcess().then((sucess) => {
                            console.log('process control In progress');
                            if (sucess) {
                                //actual method of caller
                                //console.log('method execution started', originalMethod.name, argsObj);
                                var result = originalMethod.apply(this, arguments);
                                if (Utils.isPromise(result)) {
                                    console.log('method executing...');
                                    return result.then((sucess) => {
                                        if (type == decorators_1.Decorators.PROCESS_START_AND_END) {
                                            //return statement
                                            return processControlService.completeProcess(sucess).then((result) => {
                                                console.log('method execution completed');
                                                return sucess;
                                            });
                                        }
                                        else {
                                            return sucess;
                                        }
                                    }).catch(error => {
                                        console.log("throw>>>>>>>>>>>>>>>>>>>");
                                        if (type == decorators_1.Decorators.PROCESS_START_AND_END) {
                                            //return statement
                                            return processControlService.errorOutProcess(JSON.stringify(error)).then((result) => {
                                                console.log('method execution error', error);
                                                throw error;
                                            }).catch((err) => {
                                                console.log('method execution error', error);
                                                throw error;
                                            });
                                        }
                                        else {
                                            throw error;
                                        }
                                    });
                                }
                                else {
                                    if (type == decorators_1.Decorators.PROCESS_START_AND_END) {
                                        //return statement
                                        console.log('method execution started', originalMethod.name, argsObj);
                                        return processControlService.completeProcess(result).then(res => {
                                            console.log('method execution completed');
                                            return result;
                                        });
                                    }
                                }
                            }
                            else {
                                return Promise.reject("already running process");
                            }
                        }, (error) => {
                            return Promise.reject("Error while starting process : " + error);
                        });
                    });
                }
            }
            if (type == decorators_1.Decorators.PROCESS_END) {
                var result = originalMethod.apply(this, arguments);
                if (Utils.isPromise(result)) {
                    return result.then((sucess) => {
                        return processControlService.completeProcess(sucess).then((result) => {
                            return sucess;
                        });
                    }, (error) => {
                        return processControlService.errorOutProcess(JSON.stringify(error)).then((result) => {
                            throw error;
                        }).catch((err) => {
                            throw error;
                        });
                    });
                }
                else {
                    processControlService.completeProcess(result);
                    return result;
                }
            }
            return descriptor;
        }
        else {
            return originalMethod.apply(this, arguments);
        }
    };
}
function processStart(params) {
    params = params || {};
    return function (target, propertyKey, descriptor) {
        utils_1.MetaUtils.addMetaData(target, {
            decorator: decorators_1.Decorators.PROCESS_START,
            decoratorType: decorator_type_1.DecoratorType.METHOD,
            params: params,
            propertyKey: propertyKey
        });
        var origianlmethod = descriptor.value;
        descriptor.value = preProcessHandler(params, target, propertyKey, descriptor, origianlmethod, decorators_1.Decorators.PROCESS_START);
        //descriptor.value = preProcessHandler(params, target, propertyKey, descriptor, rsProcessDecorators.PROCESS_START);
    };
}
exports.processStart = processStart;
function processEnd(params) {
    params = params || {};
    return function (target, propertyKey, descriptor) {
        utils_1.MetaUtils.addMetaData(target, {
            decorator: decorators_1.Decorators.PROCESS_END,
            decoratorType: decorator_type_1.DecoratorType.METHOD,
            params: params,
            propertyKey: propertyKey
        });
        var origianlmethod = descriptor.value;
        descriptor.value = preProcessHandler(params, target, propertyKey, descriptor, origianlmethod, decorators_1.Decorators.PROCESS_END);
        //descriptor.value = preProcessHandler(params, target, propertyKey, descriptor, rsProcessDecorators.PROCESS_END);
    };
}
exports.processEnd = processEnd;
function processStartEnd(params) {
    params = params || {};
    return function (target, propertyKey, descriptor) {
        utils_1.MetaUtils.addMetaData(target, {
            decorator: decorators_1.Decorators.PROCESS_START_AND_END,
            decoratorType: decorator_type_1.DecoratorType.METHOD,
            params: params,
            propertyKey: propertyKey
        });
        var origianlmethod = descriptor.value;
        descriptor.value = preProcessHandler(params, target, propertyKey, descriptor, origianlmethod, decorators_1.Decorators.PROCESS_START_AND_END);
    };
}
exports.processStartEnd = processStartEnd;
