// nodedata/core/constants/decorators.js
"use strict";
class Decorators {
}
Decorators.DOCUMENT = 'document';
Decorators.ENTITY = 'entity';
Decorators.FIELD = 'field';
Decorators.COLUMN = 'column';
Decorators.INJECT = 'inject';
Decorators.INJECTBYNAME = 'injectbyname';
Decorators.MANYTOMANY = 'manytomany';
Decorators.MANYTOONE = 'manytoone';
Decorators.ONETOMANY = 'onetomany';
Decorators.ONETOONE = 'onetoone';
Decorators.REPOSITORY = 'repository';
Decorators.SERVICE = 'service';
Decorators.AUTHORIZE = 'authorize';
Decorators.PREAUTHORIZE = 'preauthorize';
Decorators.POSTFILTER = 'postfilter';
Decorators.UPLOAD = 'upload';
Decorators.JSONIGNORE = 'jsonignore';
Decorators.REQUIRED = 'required';
Decorators.TRANSIENT = 'transient';
Decorators.ALLOWANONYMOUS = 'allowanonymous';
Decorators.WORKER = 'worker';
Decorators.OPTIMISTICLOCK = 'OptimisticLocking';
Decorators.PROMISABLE = 'Promisable';
Decorators.PROCESS_START = 'processStart';
Decorators.PROCESS_END = 'processEnd';
Decorators.PROCESS_START_AND_END = 'processStartEnd';
exports.Decorators = Decorators;
(function (ExportTypes) {
    ExportTypes[ExportTypes["NONE"] = 0] = "NONE";
    ExportTypes[ExportTypes["REST"] = 1] = "REST";
    ExportTypes[ExportTypes["WS"] = 2] = "WS";
    ExportTypes[ExportTypes["WS_BROAD_CAST"] = 4] = "WS_BROAD_CAST";
    ExportTypes[ExportTypes["PUB_SUB"] = 8] = "PUB_SUB";
    ExportTypes[ExportTypes["WS_WS_BROAD_CAST_PUB_SUB"] = 14] = "WS_WS_BROAD_CAST_PUB_SUB";
    ExportTypes[ExportTypes["ALL"] = 15] = "ALL";
})(exports.ExportTypes || (exports.ExportTypes = {}));
var ExportTypes = exports.ExportTypes;
exports.RelationDecorators = [Decorators.ONETOONE, Decorators.ONETOMANY, Decorators.MANYTOONE, Decorators.MANYTOMANY];
